File: model/mixture by group and maturity/instar_group_log_maturity.cpp
#include <TMB.hpp>
template<class Type> Type objective_function<Type>::operator()(){
   // Data:
   DATA_VECTOR(x);           // Log-size measurements (n).
   DATA_VECTOR(f);           // Frequency observations (n).
   DATA_VECTOR(maturity);    // Crab maturity(n).
   DATA_VECTOR(precision);   // Precision of size measurements (n).
   DATA_IVECTOR(group);      // group identifier (n). 
   DATA_INTEGER(n_instar);   // Number of instars.
   DATA_INTEGER(n_group);    // Number of groups.

   // Instar mean parameters:
   PARAMETER(mu_instar_0);                        // Mean size of the first instar.
   PARAMETER(log_increment);                      // Vector of log-scale growth increments.
   PARAMETER(log_increment_delta);                // Vector of log-scale growth increments modifiers.
   PARAMETER_VECTOR(mu_instar_group);             // Instar-group means random effect.
   PARAMETER(log_sigma_mu_instar_group);          // Log-scale error for instar-group means random effect.
   
   // Instar error parameters:
   PARAMETER(log_sigma);                          // Log-scale instar standard error by instar.
   PARAMETER_VECTOR(log_sigma_instar_group);
   PARAMETER(log_sigma_sigma_instar_group);
   
   // Instar proportion parameters:
   PARAMETER_VECTOR(mu_logit_p);                  // Instar proportions log-scale global mean parameter (n_instar-1).
   PARAMETER(log_sigma_logit_p_instar_group);     // Instar proportions log-scale global error parameter.
   PARAMETER_VECTOR(logit_p_instar_group);        // Proportion deviations by instar and group.

   // Mature crab parameters:
   PARAMETER_VECTOR(mu_instar_group_mature);      // Instar-group means random effect for mature crab.
   PARAMETER_VECTOR(mu_logit_p_mature);           // Instar proportions log-scale global mean parameter (n_instar-1).
   PARAMETER_VECTOR(logit_p_instar_group_mature); // Proportion deviations by instar and group.

   // Vector sizes:      
   int n = x.size();
   
   // Initialize log-likelihood:
   Type v = 0;
   
   // Global instar mean sizes:
   vector<Type> mu_instar(n_instar);
   mu_instar[0] = mu_instar_0;
   for (int j = 1; j < n_instar; j++){
      mu_instar[j] = mu_instar_0 + j * exp(log_increment) - 0.5 * (j-1) * j * exp(log_increment_delta); 
      //mu_instar[j] = mu_instar[j-1] + exp(log_increment) - j * exp(log_increment_delta); // Instar means.
      //DATA_IVECTOR(ado);
      //mu_instar[j] = mu_instar[j-1] + exp(log_increment) - j * exp(log_increment_delta + ado[j] * log_increment_delta_ado); // Instar means.
   }
  
   // Group-level instar sizes and standard errors:
   v -= sum(dnorm(mu_instar_group,        0, exp(log_sigma_mu_instar_group), true)); 
   v -= sum(dnorm(mu_instar_group_mature, 0, exp(log_sigma_mu_instar_group), true)); 
   v -= sum(dnorm(log_sigma_instar_group, 0, exp(log_sigma_sigma_instar_group), true)); 
   matrix<Type> mu(n_instar,n_group);
   matrix<Type> mu_mature(n_instar,n_group);
   matrix<Type> sigma(n_instar,n_group);
   for (int i = 0; i < n_group; i++){
      for (int j = 0; j < n_instar; j++){
         mu(j,i)        = mu_instar[j] + mu_instar_group[j * n_group + i];          // Immature mean sizes.
         mu_mature(j,i) = mu_instar[j] + mu_instar_group_mature[j * n_group + i];   // Mature mean sizes.
         sigma(j,i)     = exp(log_sigma + log_sigma_instar_group[j * n_group + i]); // Instar standard errors.
      }
   }   

   // Instar proportions:
   v -= sum(dnorm(logit_p_instar_group,        0, exp(log_sigma_logit_p_instar_group), true));
   v -= sum(dnorm(logit_p_instar_group_mature, 0, exp(log_sigma_logit_p_instar_group), true));
   matrix<Type> p(n_instar,n_group);
   matrix<Type> p_mature(n_instar,n_group);
   vector<Type> sum_logit_p(n_group);
   vector<Type> sum_logit_p_mature(n_group);
   matrix<Type> logit_p(n_instar-1,n_group);
   matrix<Type> logit_p_mature(n_instar-1,n_group);
   for (int i = 0; i < n_group; i++){
      sum_logit_p[i] = 0;
      sum_logit_p_mature[i] = 0;
      for (int j = 1; j < n_instar; j++){
         // Normalizing constants:
         sum_logit_p[i]        += exp(logit_p(j-1,i)); 
         sum_logit_p_mature[i] += exp(logit_p_mature(j-1,i)); 
         
         // Logit-scale effects:
         logit_p(j-1,i)        = mu_logit_p[j-1]        + logit_p_instar_group[(j-1) * n_group + i]; 
         logit_p_mature(j-1,i) = mu_logit_p_mature[j-1] + logit_p_instar_group_mature[(j-1) * n_group + i]; 
      }
   }
   for (int i = 0; i < n_group; i++){
      p(0,i)        = 1 / (1 + sum_logit_p[i]);
      p_mature(0,i) = 1 / (1 + sum_logit_p_mature[i]);
      for (int j = 1; j < n_instar; j++){
         p(j,i)        = exp(logit_p(j-1,i))        / (1 + sum_logit_p[i]); 
         p_mature(j,i) = exp(logit_p_mature(j-1,i)) / (1 + sum_logit_p_mature[i]);
      }
   }
   
   // Mixture likelihood:
   for (int i = 0; i < n; i++){ 
      Type d = 0;
      Type xlower = log(exp(x[i]) - precision[i] / 2);
      Type xupper = log(exp(x[i]) + precision[i] / 2);
      for (int j = 0; j < n_instar; j++){
         Type pp = (1-maturity[i]) * p(j,group[i])  + maturity[i] * p_mature(j,group[i]);
         Type mm = (1-maturity[i]) * mu(j,group[i]) + maturity[i] * mu_mature(j,group[i]);
         d +=  pp * (pnorm(xupper, mm, sigma(j,group[i])) - pnorm(xlower, mm, sigma(j,group[i]))); 
      }
      v -= f[i] * log(d);
   }
   
   // Export variables:
   REPORT(mu_instar);
   REPORT(mu);
   REPORT(mu_mature);
   REPORT(sigma);   
   REPORT(p);
   REPORT(p_mature);

   return v;
}

File: model/mixture by group and maturity/instar_group_log_maturity.R
rm(list = ls())
library(TMB)
library(gulf.data)
library(gulf.graphics)

source("/Users/crustacean/Desktop/gulf-snow-crab-EAFM/R/TMB utilities.R")
source("/Users/crustacean/Desktop/gulf-snow-crab-EAFM/model/mixture by group/plot.instar.group.R")

compile("instar_group_log_maturity.cpp")
dyn.load(dynlib("instar_group_log_maturity"))

BSM <- FALSE
sex      <- 2           # Crab sex.
maturity <- 0           # Crab maturity.
years    <- 2017:2020   # Define survey years.

# Define instar mean sizes:
mu_instars <- c(3.22, 4.63, 6.62, 10.5, 14.5, 20.5, 28.1, 37.8, 51.1, 68.0, 88.0)
names(mu_instars) <- 1:11
mu_instars <- mu_instars[as.character(4:10)]
   
setwd("/Users/crustacean/Desktop/gulf-snow-crab-EAFM/model/mixture by group and maturity")

# Work computer fix:
if (Sys.getenv("RSTUDIO_USER_IDENTITY") == "SuretteTJ") Sys.setenv(BINPREF = "C:/Rtools/mingw_64/bin/")

if (BSM){
   # Load BSM data:
   b <- read.csv("BSM/BSM_fem_94-98.csv")
   b <- b[which(b$maturity %in% c("0", "1")), ]
   b$maturity <- as.numeric(b$maturity)
   b$date <- as.character(date(b))
   b <- sort(b, by = c("date", "tow.number", "number"))
}else{
   # Define data:
   b <- read.scsbio(years, survey = "regular")
   b <- b[which(b$sex == sex), ]
   b <- b[which(b$carapace.width > 6), ]
   b$maturity <- is.mature(b)
   b <- b[-which(b$maturity & !is.new.shell(b)), ]
   b$year <- year(b)
   b <- sort(b, by = c("date", "tow.number", "crab.number"))
   b$carapace.width <- as.numeric(b$carapace.width)
   b <- b[!is.na(b$carapace.width), ]
}

# Create tow table and tow index:
tows  <- aggregate(list(n = b$carapace.width),  b[c("date", "tow.number")], length)
tows  <- sort(tows, by = c("date", "tow.number"))
b$tow <- match(b[c("date", "tow.number")], tows[c("date", "tow.number")]) - 1
tows  <- aggregate(list(n = b$carapace.width),  b[c("date", "tow", "tow.number")], length)
tows$n_imm <- aggregate(list(x = !b$maturity),  b[c("date", "tow.number")], sum, na.rm = TRUE)$x
tows$n_mat <- aggregate(list(x = b$maturity),  b[c("date", "tow.number")], sum, na.rm = TRUE)$x
tows  <- sort(tows, by = c("date", "tow.number"))
s <- read.scsset(year = years, valid = 1, survey = "regular")
ix <- match(tows[c("date", "tow.number")], s[c("date", "tow.number")])
tows$swept.area <- s$swept.area[ix]
   
# Set up data:
#r <- aggregate(list(f = b$year), list(x = round(log(b$carapace.width), 2), year = b$year), length)
data <- aggregate(list(f = b$year), 
                  by = list(date = b$date,
                            tow.number = b$tow.number,
                            group = b$tow, 
                            year = b$year, 
                            maturity = b$maturity,
                            x = round(log(b$carapace.width), 2)), 
                   length)
data <- as.list(data)
data$n_instar <- length(mu_instars)
data$n_group  <- max(data$group) + 1
data$precision <- rep(0.1, length(data$x))

# Define initial parameters:
parameters <- list(mu_instar_0 = as.numeric(log(mu_instars[1])),   # Mean size of the first instar.
                   log_increment = log(0.345),                         # Log-scale mean parameter associated with instar growth increments.
                   log_increment_delta = log(0.010),
                   log_sigma_mu_instar_group = -4,                 # Log-scale error for instar-group means random effect.
                   mu_instar_group = rep(0, length(mu_instars) * length(unique(data$group))),  # Instar-group means random effect.
                   mu_instar_group_mature = rep(0, length(mu_instars) * length(unique(data$group))),  # Instar-group means random effect.
                   log_sigma = -2,             # Log-scale instar standard error.
                   log_sigma_instar_group = rep(0, length(mu_instars) * length(unique(data$group))),
                   log_sigma_sigma_instar_group = -4,
                   mu_logit_p = rep(4, data$n_instar-1),           # Instar proportions log-scale mean parameter.
                   mu_logit_p_mature = rep(4, data$n_instar-1), 
                   log_sigma_logit_p_instar_group = -1,            # Instar-group proportions log-scale error parameter.
                   logit_p_instar_group = rep(0, (length(mu_instars)-1) * length(unique(data$group))), # Multi-logit-scale parameters for instar-group proportions by year.
                   logit_p_instar_group_mature = rep(0, (length(mu_instars)-1) * length(unique(data$group)))
                   ) 
  
random     <- c("mu_instar_group", "mu_instar_group_mature", "logit_p_instar_group", "logit_p_instar_group_mature", "log_sigma_instar_group")

# Fit mixture proportions for immatures:
map <- lapply(parameters, function(x) as.factor(rep(NA, length(x))))
map$mu_logit_p <- factor(1:(data$n_instar-1))
map$log_sigma_logit_p_instar_group <- factor(1)
map$logit_p_instar_group <- factor(1:length(parameters$logit_p_instar_group))
map$mu_logit_p_mature <- factor(1:(data$n_instar-1))
map$log_sigma_logit_p_instar_group <- factor(1)
map$logit_p_instar_group_mature <- factor(1:length(parameters$logit_p_instar_group))
map$log_sigma <- factor(1)

obj <- MakeADFun(data[data.cpp("instar_group_log_maturity.cpp")], 
                 parameters[parameters.cpp("instar_group_log_maturity.cpp")], 
                 random = random, map = map, DLL = "instar_group_log_maturity")
theta <- optim(obj$par, obj$fn, control = list(trace = 3, maxit = 1000))$par
obj$par <- theta
rep <- sdreport(obj)
parameters <- update.parameters(parameters, summary(rep, "fixed"), summary(rep, "random"))

          
#map$log_sigma <- factor(rep(1,data$n_instar))


map$log_increment       <- factor(1)
map$log_increment_delta <- factor(1)

map$mu_instar_0 <- factor(1)
map$mu_instar_0 <- factor(NA)

map$mu_instar_group <- factor(1:length(parameters$mu_instar_group))
map$mu_instar_group_mature <- factor(1:length(parameters$mu_instar_group_mature))

map$log_sigma_mu_instar_group <- factor(1)




